File: spec/java/src/io/kaitai/struct/specwrite/TestRepeatUntilS4.java
package io.kaitai.struct.specwrite;

import io.kaitai.struct.ConsistencyError;
import io.kaitai.struct.KaitaiStruct.ReadWrite;
import io.kaitai.struct.testwrite.RepeatUntilS4;

import java.util.ArrayList;
import java.util.Arrays;

import org.testng.annotations.Test;

public class TestRepeatUntilS4 extends CommonSpec {
    @Override
    protected Class<? extends ReadWrite> getStructClass() {
        return RepeatUntilS4.class;
    }

    @Override
    protected String getSrcFilename() {
        return "repeat_until_s4.bin";
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: entries,.*")
    public void testCheckBadNoEntries() {
        RepeatUntilS4 r = new RepeatUntilS4();
        r.setEntries(new ArrayList<>(0));
        r._check();
    }

    @Test
    public void testCheckGoodOneEntry() {
        RepeatUntilS4 r = new RepeatUntilS4();
        r.setEntries(new ArrayList<>(Arrays.asList(-1)));
        r.setAfterall("");
        r._check();
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: entries,.*")
    public void testCheckBadEarlyUntilEntry() {
        RepeatUntilS4 r = new RepeatUntilS4();
        r.setEntries(new ArrayList<>(Arrays.asList(-3, 275000, -1, 0, -1)));
        r._check();
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: entries,.*")
    public void testCheckBadNoUntilEntry() {
        RepeatUntilS4 r = new RepeatUntilS4();
        r.setEntries(new ArrayList<>(Arrays.asList(-3, 275000, -2, 0)));
        r._check();
    }

    @Test
    public void testCheckGoodUntilEntry() {
        RepeatUntilS4 r = new RepeatUntilS4();
        r.setEntries(new ArrayList<>(Arrays.asList(-3, 275000, -2, 0, -1)));
        r.setAfterall("");
        r._check();
    }
}

File: spec/python/specwrite/test_repeat_until_s4.py
import unittest
from kaitaistruct import ConsistencyError
from specwrite.common_spec import CommonSpec

from testwrite.repeat_until_s4 import RepeatUntilS4

class TestRepeatUntilS4(CommonSpec.Base):
    def __init__(self, *args, **kwargs):
        super(TestRepeatUntilS4, self).__init__(*args, **kwargs)
        self.struct_class = RepeatUntilS4
        self.src_filename = 'src/repeat_until_s4.bin'

    def test_check_bad_no_entries(self):
        r = RepeatUntilS4()
        r.entries = [0]
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: entries,"):
            r._check()

    def test_check_good_one_entry(self):
        r = RepeatUntilS4()
        r.entries = [-1]
        r.afterall = ""
        r._check()

    def test_check_bad_early_until_entry(self):
        r = RepeatUntilS4()
        r.entries = [-3, 275000, -1, 0, -1]
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: entries,"):
            r._check()

    def test_check_bad_no_until_entry(self):
        r = RepeatUntilS4()
        r.entries = [-3, 275000, -2, 0]
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: entries,"):
            r._check()

    def test_check_good_until_entry(self):
        r = RepeatUntilS4()
        r.entries = [-3, 275000, -2, 0, -1]
        r.afterall = ""
        r._check()
